platform/ios/src/MGLMapboxEvents.m

```diff
@@ -74,60 +76,64 @@ - (instancetype)init {
             self.baseURL = [NSURL URLWithString:[[NSUserDefaults standardUserDefaults] objectForKey:MGLTelemetryBaseURLKey]];
         }
         
-        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(userDefaultsDidChange:) name:NSUserDefaultsDidChangeNotification object:nil];
+        // Guard against over calling pause / resume if the values this implementation actually
+        // cares about have not changed. We guard because the pause and resume method checks CoreLocation's
+        // authorization status and that can drag on the main thread if done too many times (e.g. if the host
+        // app heavily uses the user defaults API and this method is called very frequently)
+        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+        [defaults addObserver:self
+                   forKeyPath:MGLMapboxMetricsEnabledKey
+                      options:NSKeyValueObservingOptionNew
+                      context:MGLMapboxMetricsEnabledKeyContext];
+        [defaults addObserver:self
+                   forKeyPath:MGLMapboxMetricsDebugLoggingEnabledKey
+                      options:NSKeyValueObservingOptionNew
+                      context:MGLMapboxMetricsDebugLoggingEnabledKeyContext];
+        [defaults addObserver:self
+                   forKeyPath:MGLTelemetryAccessTokenKey
+                      options:NSKeyValueObservingOptionNew
+                      context:MGLTelemetryAccessTokenKeyContext];
     }
     return self;
 }
 
 - (void)dealloc {
-    [[NSNotificationCenter defaultCenter] removeObserver:self];
+    @try {
+        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+        [defaults removeObserver:self forKeyPath:MGLMapboxMetricsEnabledKey];
+    }
+    @catch (NSException *exception) {} //Purposefully unhandled. This should only fail if we are unable to add an observer in the first place.
 }
```

Comment:
It's unlikely but if another class were to observe this same parameter and removed the observer and then we tried to remove it a second time then it could cause a crash. Since we're setting this at init and only use these keys within this class; I don't think it should but I just wanted to add it as a precaution (even if it is an unnecessary one).

Reply:
Previous code comment probably wouldn't happen. I cleaned the comment up with the more realistic scenario as well.